### PR TITLE
V14 dingo swap cap actions

### PR DESCRIPTION
**Pull request by aljones15**
This changes all proofs with capabilityAction to `write`.

Sister PRs:
https://github.com/veres-one/veres-one/pull/49
https://github.com/veres-one/veres-one-validator/pull/35

### CODE REVIEW

**Review by dmitrizagidulin — APPROVED** [review]
:+1: 
Looks reasonable

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,9 +3,12 @@
 ## 14.0.0 -
 
 ### Changed
-- BREAKING: Replaced axios with @digitalbazaar/http-client. Errors returned directly from http-client do not match the axios API.
+- **BREAKING**: Replaced axios with @digitalbazaar/http-client. Errors returned directly from http-client do not match the axios API.
 - Changed API `getTicketServiceProof` to use `@digitalbazaar/http-client`.
 - Changed API `sendToAccelerator` to use `@digitalbazaar/http-client`.
+- **BREAKING**: API attachInvocationProof now requires the parameter `invocationTarget`.
+- **BREAKING**: API attachInvocationProof now accepts a single object as parameters.
+
```
Comment (dlongley) [inline CHANGELOG.md:11]:
```suggestion
```

File: lib/attachProof.js [hunk 2]
```diff
@@ -39,12 +39,14 @@ const {Ed25519Signature2020} = require('@digitalbazaar/ed25519-signature-2020');
  * @param {object} [options.authDoc] - Auth DID Doc, required if using
  *   an accelerator service.
  * @param {string} options.mode - Ledger mode ('test', 'live' etc).
+ * @param {string} options.invocationTarget - The invocationTarget
+ *   (should be the ledger id or the id of a record on the ledger).
```
Comment (dlongley) [inline lib/attachProof.js:43]: I think this should be removed. The `invocationTarget` here should *always* be the DID identified by the record ID/recordPatch target. The caller should not have to specify it.

The accelerator will use a different invocation target, but they will be responsible for that -- and theirs will always be the same as well -- the ID for the veres one ledger.

```suggestion
```

File: lib/attachProof.js [hunk 3]
```diff
@@ -39,12 +39,14 @@ const {Ed25519Signature2020} = require('@digitalbazaar/ed25519-signature-2020');
...
  *
  * @returns {Promise<object>} - An operation document with proofs attached.
  */
 async function attachProofs(operation, {
   did, client, capabilityInvocationKeyPair = {}, signer = {}, logger,
-  authenticationKeyPair, authDoc, accelerator, mode
+  authenticationKeyPair, authDoc, accelerator, mode, invocationTarget
```
Comment (dlongley) [inline lib/attachProof.js:49]:
```suggestion
  authenticationKeyPair, authDoc, accelerator, mode
```

File: lib/attachProof.js [hunk 4]
```diff
@@ -57,15 +59,18 @@ async function attachProofs(operation, {
     // send to ticket service for a proof
     operation = await attachTicketServiceProof({client, ...operation});
   }
-
+  // if there is no invocationTarget defined try to find one
+  if(!invocationTarget) {
+    invocationTarget = _getInvocationTarget({operation});
+  }
```
Comment (dlongley) [inline lib/attachProof.js:65]:
```suggestion
  // get invocation target from the record/recordPatch in the operation
  const invocationTarget = _getInvocationTarget({operation});
```

File: lib/attachProof.js [hunk 5]
```diff
@@ -57,15 +59,18 @@ async function attachProofs(operation, {
...
-    'create' : 'write';
-  operation = await attachInvocationProof(operation, {
+  operation = await attachInvocationProof({
+    operation,
     capability: did,
-    capabilityAction,
+    capabilityAction: 'write',
     controller,
+    invocationTarget,
     key: capabilityInvocationKeyPair,
```
Comment (dlongley) [inline lib/attachProof.js:74]: Why is a `key` being passed in addition to a `signer`? I see that's what was happening before and this PR doesn't change it, but it seems unnecessary (only a signer should be needed). Just noting that.

File: lib/attachProof.js [hunk 6]
```diff
@@ -139,6 +154,16 @@ async function attachTicketServiceProof({client, ...operation} = {}) {
...
 }
 
+function _getInvocationTarget({operation}) {
+  if(operation.record) {
+    return operation.record.id;
+  }
+  if(operation.recordPatch) {
+    return operation.recordPatch.target;
+  }
+  throw new Error('Unable to determine invocationTarget for record.');
```
Comment (dlongley) [inline lib/attachProof.js:164]:
```suggestion
  throw new Error('Unable to determine invocationTarget for operation.');
```

**Review by dlongley — APPROVED** [review]
Approving on the condition that my suggestions are applied. Thanks!

File: lib/attachProof.js (hunk 5 above)
Comment (aljones15) [reply lib/attachProof.js:70]: @dmitrizagidulin any idea on this one?

Comment (msporny) [reply lib/attachProof.js:70]: If key is passed in, signer is derived from it. I agree that we should just be passing in a signer... a lot of the ?legacy? code tries to pass the key in. There are other issues w/ the design, such as naming this thing a VerificationKey in a lot of places when it's really a keypair... we need to have a redesign discussion after we make the latest releases... I think the serialized structures are fine for now... but the library design needs a bit of work.

File: lib/attachProof.js [hunk 7]
```diff
@@ -57,15 +57,16 @@ async function attachProofs(operation, {
...
   const controller = capabilityInvocationKeyPair.id || signer.id;
   // attach capability invocation proof
-  const capabilityAction = (operation.type === 'CreateWebLedgerRecord') ?
-    'create' : 'write';
-  operation = await attachInvocationProof(operation, {
+  operation = await attachInvocationProof({
+    operation,
     capability: did,
-    capabilityAction,
+    capabilityAction: 'write',
```
Comment (msporny) [inline lib/attachProof.js:67]: Isn't write the default?

File: lib/attachProof.js [hunk 8]
```diff
@@ -120,13 +122,22 @@ async function attachAcceleratorProof(operation, {
  *
  * @returns {Promise<object>}
  */
-async function attachInvocationProof(operation, {
-  capability, capabilityAction, key, signer
+async function attachInvocationProof({
+  operation,
+  capability,
+  capabilityAction = 'write',
```
Comment (msporny) [inline lib/attachProof.js:128]: is it dangerous to have 'write' as the default? Perhaps we shouldn't have a default here? Then again, there is no other useful choice for 'attach', is there?

File: lib/attachProof.js (hunk 5 above)
Comment (aljones15) [reply lib/attachProof.js:70]: I made an issue about it: https://github.com/veres-one/did-veres-one/issues/70

**Review by msporny — APPROVED** [review]

File: lib/attachProof.js (hunk 8 above)
Comment (dlongley) [reply lib/attachProof.js:128]: There is no other capability action for veres one, I'd say remove the param.

**Pull request merged**